Add generic isnull and unsafe_get methods

davidagold opened the pull request:
It would be useful to have a generic fallback for `isnull` when operating on iterators (usually tuples) of both `Nullable` and non-`Nullable` objects. Is there anything else we need to do here to make sure this is optimized away in the case of non-`Nullable` arguments? What should tests look like?

cc: @johnmyleswhite @nalimilan @quinnj

davidagold committed:
Add generic isnull, unsafe_get methods

Also move isnull docstring out of helpdb and into base/nullable.jl.

johnmyleswhite commented:
I'm not sure how Base wants to handle things, but I would say that the removal of a dead branch in code like the following is what we require:

```julia
function loop(x::Vector)
    s = 0.0
    for x_i in x
        if !isnull(x_i)
            s += x_i
        end
    end
    return s
end

code_llvm(loop, Vector{Float64})
```

davidagold commented:
That makes sense. I will leave that assessment to somebody who can read LLVM code.

This PR also seems an appropriate place to introduce an analogous method for the `value` field of a `Nullable`:

```julia
unwrap(x::Nullable) = x.value
unwrap(x) = x
```

as in John's jplyr notes.

johnmyleswhite commented:
I think we definitely want something like `unwrap`, although we might want to call it something like `unsafe_get` since it is essentially a version of `get(x::Nullable)` that is unsafe, but extended to work on non-nullables (in a way that ironically ends up being safe).

nalimilan commented:
Regarding `unsafe_unwrap`, maybe rename `get` to `unwrap` by the way?

eschnett commented:
In other contexts (e.g. C++ futures), `unwrap` is a slightly different operation: It converts `Nullable{Nullable{T}}` to `Nullable{T}`, which is always safe. That is, it peels off one layer if there are multiple layers.

TotalVerb commented:
LLVM can already eliminate the branch, but inference cannot. See Dead branch elimination after inlining https://github.com/JuliaLang/julia/issues/17880

davidagold commented:
Should this wait for or become part of https://github.com/JuliaLang/julia/pull/18510 ?

nalimilan commented:
> Should this wait for or become part of #18510 ?

The two changes can be made independently, so we can merge any of them we want.

nalimilan commented:
@davidagold Should probably add a test for the new method? Then I vote for merging.

davidagold commented:
Ready for review.

nalimilan commented on test/nullable.jl:
```diff
@@ -161,6 +161,33 @@ for T in types
+
+@test_throws UndefRefError unsafe_get(Nullable())
+
+# unsafe_get(x)
```
Could as well merge this loop with the one above.

davidagold replied:
I suppose? But it doesn't need to be executed multiple times.

nalimilan commented on test/nullable.jl:
```diff
@@ -161,6 +161,33 @@ for T in types
+    @test unsafe_get(x4) === a
+end
+
+@test_throws UndefRefError unsafe_get(Nullable())
```
Maybe also test with another type, e.g. `Nullable{String}()`?

nalimilan commented on base/nullable.jl:
```diff
@@ -61,7 +61,11 @@ end
 
 get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
 
+unsafe_get(x::Nullable) = x.value
```
Needs a docstring.

nalimilan commented on test/nullable.jl:
```diff
@@ -174,6 +201,17 @@ end
 @test isnull(Nullable())
 
+# isnull(x)
+for T in types
```
Same here.

davidagold commented:
I've added a docstring for `unsafe_get` added, modified the `isnull` docstring, and moved the latter from `docs/helpdb/Base.jl` to `base/nullable.jl`. I've also incorporated Milan's suggestions.

nalimilan commented on base/nullable.jl:
```diff
@@ -61,7 +61,25 @@ end
+
+Return the value of `x` for `x::Nullable`; return `x` for all other `x`.
+
+Unsafe because does not check whether or not `x` is null before attempting to
```
Maybe "This function is unsafe because it does not..."?

davidagold commented:
@nalimilan How's this?

nalimilan commented on base/nullable.jl:
```diff
@@ -61,6 +61,68 @@ end
+"""
+    unsafe_get(x)
+
+Return the value of `x` for [`Nullable`](:obj:`Nullable`) `x`; return `x` for
```
Just being curious, do we need this to get links?

davidagold replied:
That's my understanding. @kshyatt included these in https://github.com/JuliaLang/julia/pull/18511/files#diff-1955b2684af6e5a407aa36281f8bafdeR94, so I figured I'd include them here, too.

nalimilan replied:
Is this documented somewhere? Maybe somebody could add a word about this at http://docs.julialang.org/en/latest/manual/documentation/? @MichaelHatherly

(FWIW, the repetition of the object name sounds a bit annoying to me...)

MichaelHatherly replied:
That syntax is being removed in #18588 in favour of `` [`Nullable`](@ref) ``, so probably no need to add it to the docs at this point.

> so I figured I'd include them here, too.

We can never have too many cross references :+1:

davidagold replied:
Okay, I'll remove the references for now. I think we will all live with the repetition in the docstring.

tkelman replied:
does the auto conversion script translate them from the current rst-in-markdown hybrid to the new style? if so it won't hurt to add if this gets merged before the conversion

MichaelHatherly replied:
Yes, handled automatically. It's fine to keep adding xrefs.

davidagold replied:
Okay, I will add it back in.

davidagold commented:
I think everything is good here. @nalimilan ?

tkelman commented:
These docstrings need to be added to the manual if this is exported

tkelman commented:
This was also missing a `make docs` update to the modified docstring. That should be done at the same time, or it will end up in someone else's unrelated PR.

tkelman commented on base/nullable.jl:
```diff
@@ -61,6 +61,68 @@ end
+isnull(x::Nullable) = x.isnull
+isnull(x) = false
+
 isnull(x::Nullable) = !x.hasvalue
```
bad rebase

davidagold commented:
Yeesh, who knew I was such a scrub. I'll fix this and the documentation bit
today.

On Monday, October 3, 2016, Tony Kelman <notifications@github.com> wrote:

> *@tkelman* commented on this pull request.
> ------------------------------
>
> In base/nullable.jl
> <https://github.com/JuliaLang/julia/pull/18484#pullrequestreview-2477207>:
>
> > +julia> x = Nullable(1, true)
> +Nullable{Int64}()
> +
> +julia> isnull(x)
> +true
> +
> +julia> x = 1
> +1
> +
> +julia> isnull(x)
> +false
> +```
> +"""
> +isnull(x::Nullable) = x.isnull
> +isnull(x) = false
> +
>  isnull(x::Nullable) = !x.hasvalue
>
> bad rebase
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/18484#pullrequestreview-2477207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALUCm5Bkx2EbY7dqHH6AKbyzmRrXpEGsks5qwLA_gaJpZM4J7-CG>
> .
>

davidagold commented:
@tkelman should I revert this or put those fixes in a new PR?

tkelman commented:
little cleanup in a new pr is all that's needed, it's not that bad
